Sources/Modules/Yggdrasil/YggdrasilTown/NPCs/RockElemental.cs

```diff
@@ -173,214 +220,540 @@ public override void AI()
+						RandomState();
 					}
 					break;
 				}
 		}
 
+		// 防御状态下受击特效计时器
+		if (NPC.ai[1] > 0)
+		{
+			NPC.ai[1] -= 0.5f;
+		}
+		else
+		{
+			NPC.ai[1] = 0f;
+		}
+	}
+
+	public void StartToSuckThrowingStone()
+	{
+		Player player = Main.player[NPC.target];
+		SuckPoint = GetClosestTilePos(player.Center + new Vector2(0, -200));
+		if (SuckPoint != Vector2.Zero)
+		{
+			NextStandPoint = SuckPoint + GetNormalOfTile(SuckPoint) * 150f;
+			State = (int)BehaviorState.Throw;
+			NPC.ai[0] = 400;
+		}
 	}
-	private static Terraria.WorldBuilding.Conditions.NotNull _cachedConditions_notNull = new Terraria.WorldBuilding.Conditions.NotNull();
-	private static Terraria.WorldBuilding.Conditions.IsSolid _cachedConditions_solid = new Terraria.WorldBuilding.Conditions.IsSolid();
+
+	/// <summary>
+	/// 随机AI
+	/// </summary>
+	public void RandomState()
+	{
+		NPC.ai[3] = 0;
+		NPC.ai[2] = 0;
+		if (NPC.rotation > MathHelper.Pi)
+		{
+			NPC.rotation -= MathHelper.TwoPi;
+		}
+		if (NPC.rotation < -MathHelper.Pi)
+		{
+			NPC.rotation += MathHelper.TwoPi;
+		}
+		switch (Main.rand.Next(3))
+		{
+			case 0:
+				StartToSuckThrowingStone();
+				break;
+			case 1:
+				State = (int)BehaviorState.Dash;
+				NPC.ai[0] = 0;
+				NPC.frameCounter = 0;
+				break;
+			case 2:
+				SoundEngine.PlaySound(SoundID.DD2_BetsysWrathShot, NPC.Center);
+				State = (int)BehaviorState.Move;
+				break;
+		}
+	}
+
+	/// <summary>
+	/// 碰撞
+	/// </summary>
+	public void Collide()
+	{
+		bool collide = false;
+		if (Collision.SolidCollision(NPC.position + new Vector2(NPC.velocity.X, 0), NPC.width, NPC.height))
+		{
+			NPC.velocity.X *= -0.9f;
+			NPC.position += NPC.velocity;
+			collide = true;
+		}
+		if (Collision.SolidCollision(NPC.position + new Vector2(0, NPC.velocity.Y), NPC.width, NPC.height))
+		{
+			NPC.velocity.Y *= -0.9f;
+			NPC.position += NPC.velocity;
+			collide = true;
+		}
+		foreach (NPC npc in Main.npc)
+		{
+			if (npc != null && npc.active && npc != NPC)
+			{
+				if (npc.type == Type)
+				{
+					if (Rectangle.Intersect(npc.Hitbox, NPC.Hitbox) != Rectangle.emptyRectangle)
+					{
+						npc.velocity += Utils.SafeNormalize(npc.Center - NPC.Center, Vector2.zeroVector) * 4f;
+						npc.velocity += NPC.velocity * 0.5f;
+						if (npc.velocity.Length() > 20f)
+						{
+							npc.velocity = npc.velocity / npc.velocity.Length() * 20f;
+						}
+						npc.position += npc.velocity;
+						NPC.velocity -= Utils.SafeNormalize(npc.Center - NPC.Center, Vector2.zeroVector) * 4f;
+						NPC.velocity += npc.velocity * 0.5f;
+						if (NPC.velocity.Length() > 20f)
+						{
+							NPC.velocity = NPC.velocity / NPC.velocity.Length() * 20f;
+						}
+						NPC.position += NPC.velocity;
+						collide = true;
+					}
+				}
+			}
+		}
+		if (collide)
+		{
+			SoundEngine.PlaySound(SoundID.NPCHit2.WithPitchOffset(0.5f).WithVolumeScale(0.5f), NPC.Center);
+		}
```

Comment:
The collide method should be improved to avoid the case where the rock element keeps colliding with blocks and playing the hit sound making the player annoyed. Additionally, i think the rock element should actively crawl out instead of getting stuck in the middle if it is spawned within a block wall.